src/digital_input_stamper.rs

```diff
@@ -0,0 +1,146 @@
+///! either directly read from the timer channel or can be collected asynchronously via DMA
+///! collection.
+///!
+///! When DMA is used for timestamp collection, a DMA transfer is configured to collect as many
+///! timestamps as there are samples, but it is intended that this DMA transfer should never
+///! complete. Instead, when all samples are collected, the module pauses the DMA transfer and
+///! checks to see how many timestamps were collected. These collected timestamps are then returned
+///! for further processing.
+///!
+///! To prevent silently discarding timestamps, the TIm5 input capture over-capture interrupt is
+///! used. Any over-capture event (which indicates an overwritten timestamp) then generates an ISR
+///! which handles the over-capture.
+///!
+///! # Tradeoffs
+///! It appears that DMA transfers can take a significant amount of time to disable (400ns) if they
+///! are being prematurely stopped (such is the case here). As such, for a sample batch size of 1,
+///! this can take up a significant amount of the total available processing time for the samples.
+///! To avoid this, the module does not use DMA when the sample batch size is one. Instead, the
+///! module manually checks for any captured timestamps from the timer capture channel manually. In
+///! this mode, the maximum input clock frequency supported is equal to the configured sample rate.
+///!
+///! There is a small window while the DMA buffers are swapped where a timestamp could potentially
+///! be lost. To prevent this, the `acuire_buffer()` method should not be pre-empted. Any lost
+///! timestamp will trigger an over-capture interrupt.
+use super::{
+    hal, timers, DmaConfig, PeripheralToMemory, Transfer, SAMPLE_BUFFER_SIZE,
+};
+
+// The DMA buffers must exist in a location where DMA can access. By default, RAM uses DTCM, which
+// is off-limits to the normal DMA peripheral. Instead, we use AXISRAM.
+#[link_section = ".axisram.buffers"]
+static mut BUF: [[u32; SAMPLE_BUFFER_SIZE]; 2] = [[0; SAMPLE_BUFFER_SIZE]; 2];
+
+/// The timestamper for DI0 reference clock inputs.
+pub struct InputStamper {
+    _di0_trigger: hal::gpio::gpioa::PA3<hal::gpio::Alternate<hal::gpio::AF2>>,
+    next_buffer: Option<&'static mut [u32; SAMPLE_BUFFER_SIZE]>,
+    transfer: Option<
+        Transfer<
+            hal::dma::dma::Stream6<hal::stm32::DMA1>,
+            timers::tim5::Channel4InputCapture,
+            PeripheralToMemory,
+            &'static mut [u32; SAMPLE_BUFFER_SIZE],
+        >,
+    >,
+    capture_channel: Option<timers::tim5::Channel4InputCapture>,
+}
+
+impl InputStamper {
+    /// Construct the DI0 input timestamper.
+    ///
+    /// # Args
+    /// * `trigger` - The capture trigger input pin.
+    /// * `stream` - The DMA stream to use for collecting timestamps.
+    /// * `timer_channel - The timer channel used for capturing timestamps.
+    /// * `batch_size` - The number of samples collected per processing batch.
```

Comment:
Probably better to call this 
```suggestion
    /// * `buffer_size` - The maximum number of timestamps collected per processing batch.
```

to distinguish between the (variable) number of timestamps in the buffer and (fixed) number of ADC/DAC samples in the batch.

It makes sense to be able to force the non-DMA mode even for larger batches.

Reply:
Is it still appropriate to use the samples per batch as an upper-bound on the number of timestamps? We still use `SAMPLE_BUFFER_SIZE` to declare the size of the DMA transfer (and thus the maximum buffer_size).

Reply:
As discussed I would actually lean towards ripping out the DMA here and just say that there will be max 1 timestamp per batch.